### ISSUE TITLE
Submit new tracks to last.fm

### ISSUE THREAD

**Issue by anthonypetrov** [david-sabata/web-scrobbler#452]
Sometimes a track cannot be recognized automatically. The extension's icon is gray then. Clicking the icon I can open a popup window where I can enter the Artist and Track names manually and try to Submit them. However, the extension still checks with the last.fm database and won't let me submit new tracks.

I actually would like to be able to submit new tracks to the last.fm database.

Technically, this is possible. The track.scrobble request should only add the "duration" key (in addition to the "artist" and "track" keys), and then last.fm will happily add the new track to its database.

**Comment by inverse** [comment]
@anthonypetrov Did you get any further with your patches? Willing to share?

**Comment by anthonypetrov** [comment]
@david-sabata any update on this? Any estimate as to when these changes could be incorporated into the mainline? The change is so simple - just pass the detected track length to last.fm, and it will allow you to add unrecognized tracks...

@inverse they still work for me. However, Chrome now enforces extensions content verification, so I'm no longer able to patch the extension in place. I had to remove it from Chrome (goodbye updates to last.fm scrobbler... :( ), make a copy of the patched extension folder on my local disk, enable the Developer Mode in Chrome, and then load the unpacked extension from the copy folder. It works just fine now.

As for sharing the patches, sure! But this bug tracking system won't let me attach patches. It only supports images. I could send them to you by email if I had one. Or I can paste them in a comment, but I'm afraid the formatting/spacing could suffer so that you'd have to apply them manually (currently they apply cleanly to version 1.43). Let me know what option works best for you.

**Comment by inverse** [comment]
@anthonypetrov  Are you able to put your changes to a branch and share that?

**Comment by anthonypetrov** [comment]
I could, but have no time. Here's the patches:

```
--- popup.js.orig	2014-11-09 16:23:57.761325200 -0800
+++ popup.js	2014-11-09 16:27:39.178123400 -0800
@@ -87,10 +87,20 @@
 
 			var validateCB = function (res) {
 				if (res === false) {
-					$(this).attr('disabled', false);
-					$(this).siblings().remove();
-					$(this).parent().append('<span class="note">Not valid</span>');
-				} else {
+                    var ar = artist;
+                    var tr = track;
+                    var dr = 60;
+                    if (!song.duration) {
+                        dr = 60;
+                    } else {
+                        dr = song.duration;
+                    }
+                    song.submitDuration = true;
+                    res =  {artist : ar,
+                        track : tr,
+                        duration : dr
+                    };
+                }  {
 					// did the content script recognize at least the duration?
 					if (!song.duration) {
 						song.duration = res.duration;
@@ -107,7 +117,7 @@
 					// The minimum time is 240 seconds or half the track's total length
 					// minus the time that already past
 					var past = parseInt(new Date().getTime() / 1000.0) - song.startTime;
-					var min_time = Math.min(240, song.duration / 2) - past;
+					var min_time = Math.min(30, song.duration / 5) - past;
 
 					// Set up the timer to scrobble
 					popupApi.planSubmit(min_time * 1000);
```



******************************************************************************


```
--- scrobbler.js.orig	2014-12-08 22:47:46.000000000 -0800
+++ scrobbler.js	2014-12-23 20:25:25.944198600 -0800
@@ -303,7 +303,7 @@
 			if (typeof song.album !== 'undefined' && song.album !== null) {
 				params.album = song.album;
 			}
-			if (typeof song.duration !== 'undefined' && song.duration !== null) {
+			if (!(!song.submitDuration) || (typeof song.duration !== 'undefined' && song.duration !== null)) {
 				params.duration = song.duration;
 			}
 
@@ -389,6 +389,10 @@
 			if (typeof song.sourceId !== 'undefined' && song.sourceId !== null) {
 				params['sourceId[0]'] = song.sourceId;
 			}
+            
+            if (!(!song.submitDuration)) {
+                params["duration[0]"] = song.duration;
+            }
 
 			var api_sig = LastFM.generateSign(params);
 			var url = config.apiURL + createQueryString(params) + '&api_sig=' + api_sig;
@@ -529,7 +533,7 @@
 					// The minimum time is 240 seconds or half the
 					// track's total length. Subtract the song's
 					// current time (for the case of unpausing).
-					var min_time = (Math.max(1, Math.min(240, song.duration / 2) - song.currentTime));
+					var min_time = (Math.max(1, Math.min(30, song.duration / 5) - song.currentTime));
 					// Set up the timer
 					scrobbleTimeout = setTimeout(submit, min_time * 1000);
 				}
```


Hope this helps.

**Comment by inverse** [comment]
Thanks for the share! so basically sets the duration to 60 for when it is not defined?

Perhaps a more elegant solution would be to:

-  Allow the popup to be populated with what the scrobbler believes to be a new track is.
-  Include a duration field that if the scrobbler supports duration handling to set that also, or allow the user to supply this with perhaps some sort formatting applied e.g. `mm:ss`

Opinions?

**Comment by anthonypetrov** [comment]
In 99.9% of cases it can detect the correct track length, so I didn't feel like it's necessary to implement any advanced logic/UI for that. But you and/or @david-sabata are free to improve this part.

The crux of the patch is 1) to avoid the "Not valid" state in the extension, AND 2) pass the duration of the track to last.fm when the scrobble is submitted.

(The changes to the minimum duration of the song before it gets submitted are unrelated, but they're a part of my patches, so I left them as they are. I've filed a separate feature request for that: #453 ).

**Comment by david-sabata** [comment]
@inverse the `song` object within `scrobbler.js` is much like singleton. It's created when the extension loads and then only its properties are changed. That's why `popup.js` directly writes to it and it works.

Also the popup should already be filled with values that were parsed from current page, if any.

**Comment by inverse** [comment]
@david-sabata I'm not seeing this working with any `v1` scrobblers.

Example: https://soundcloud.com/eyesix/dyed-in-the-wool

Are you able the shed some light when you have some time?

**Comment by david-sabata** [comment]
@inverse it's possible that it really did stopp working some time ago. Also it cannot work if a connector doesn't send all the info it has. For example if there's a condition like `if (artist==null || track==null) { /* dont send to background */ }`, then the background script has nothing to prefill.

I probably said that somewhere else, but I'll repeat. I'm not a fan of adding new stuff for `v1` connectors or `legacy/scrobble.js`. It works "good enough" and I'd like to keep it that way rather than broke it by adding new features :)

For `v2` I'd like to have this as a part of a new popup window which will also contain song info, probably some buttons ("don't scrobble this",...) and such. But it's still in just an idea without some specific interface.

**Comment by lacivert** [comment]
@david-sabata :+1:

**Comment by janbaykara** [comment]
@david-sabata I may be wrong, but isn't this already the behaviour of the old (current) manual_scrobble popup, in addition to the upcoming new popup? Think this can be closed.

**Comment by david-sabata** [comment]
@janbaykara we only submit the track [if it is already on L.FM or has been corrected by user](https://github.com/david-sabata/web-scrobbler/blob/master/core/background/controller.js#L192). But I suppose we can drop the second condition with the new popup. It was there because user couldn't edit the info once the track was (badly) recognized

**Comment by janbaykara** [comment]
Ah apologies, I misread the OP.

**Comment by TheQwerty** [comment]
> we only submit the track if it is already on L.FM or has been corrected by user. But I suppose we can drop the second condition with the new popup.

@david-sabata Any chance the first half of this is going to become optional in v2?

I listen to a lot of new music on Google Play Music and it's rather annoying that over half of the tracks aren't scrobbling because track.getInfo returns a 400 with error code 6 "Track not found". (Meanwhile, requesting track.getInfo?format=json at least returns an HTTP 200, with the same error & message; erm, thanks Last.fm?)

This means with v1.52 if I want the tracks to scobble I have to sit here baby-sitting the extension and clicking through the unrecognized track/manual scrobble popup on just about every song.

**Comment by alexesprit** [comment]
Closed as a duplicate of #185.
